File: src/main/java/io/tripled/social/client/domain/SocialNetwork.java
package io.tripled.social.client.domain;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class SocialNetwork {

  private final Messages messages;
  private final Relationships relationships;
  private final DateTimeProvider dateTimeProvider;
  private static final MessageFormatter READING_FRMT = MessageFormatter.ofPattern("%msg (%t)");
  private static final MessageFormatter WALL_FRMT = MessageFormatter.ofPattern("%un - %msg (%t)");

  public SocialNetwork(Messages messages, Relationships relationships, DateTimeProvider dateTimeProvider) {
    this.messages = messages;
    this.relationships = relationships;
    this.dateTimeProvider = dateTimeProvider;
  }

  public void postMessage(Message message) {
    messages.save(message);
  }

  public String read(UserName userName) {
    List<Message> messagesToPrint = messages.findAllByUserName(userName);

    return READING_FRMT.print(messagesToPrint, dateTimeProvider);
  }

  public void follow(UserName userName, UserName userNameToFollow) {
    relationships.save(new FollowingRelationship(userName, userNameToFollow));
  }

  public String readWall(UserName userName) {
    List<UserName> userNames = new ArrayList<>();
    userNames.add(userName);
    userNames.addAll(relationships.findRelationshipsFor(userName).stream()
        .map(FollowingRelationship::getFollowing)
        .collect(Collectors.toList()));

    List<Message> messagesToPrint = messages.findAllByUserNames(userNames);
    return WALL_FRMT.print(messagesToPrint, dateTimeProvider);
  }
}

File: src/main/kotlin/io/tripled/social/client/domain/SocialNetwork.kt
package io.tripled.social.client.domain

class SocialNetwork(
    private val messages: Messages,
    private val relationships: Relationships,
    private val dateTimeProvider: DateTimeProvider
) {
    companion object {
        private val READING_PRINTER = MessagePrinter.ofPattern("%msg (%t)")
        private val WALL_PRINTER = MessagePrinter.ofPattern("%un - %msg (%t)")
    }

    fun postMessage(message: Message): String {
        messages.save(message)
        return "${message.getUserName()} posted: ${message.getMessage()}"
    }

    fun read(userName: UserName): String {
        val messagesToPrint = messages.findAllByUserName(userName)
        return printAll(messagesToPrint, READING_PRINTER)
    }

    fun follow(userName: UserName, userNameToFollow: UserName): String {
        relationships.save(FollowingRelationship(userName, userNameToFollow))
        return "$userName will follow $userNameToFollow"
    }

    fun readWall(userName: UserName): String {
        val userNames = mutableListOf<UserName>().apply {
            add(userName)
            addAll(relationships.findRelationshipsFor(userName).map { it.getFollowing() })
        }

        val messagesToPrint = messages.findAllByUserNames(userNames)
        return printAll(messagesToPrint, WALL_PRINTER)
    }

    private fun printAll(messagesToPrint: List<Message>, printer: MessagePrinter): String =
        messagesToPrint
            .joinToString(System.lineSeparator()) { msg ->
                msg.printWith(printer, dateTimeProvider)
            }
            .trim()
}